### ISSUE TITLE
Shorten container status and health to 1 character

### ISSUE THREAD

**Issue by martini1992** [jesseduffield/lazydocker#404]
**Is your feature request related to a problem? Please describe.**
Using the container list can be awkward when forces to use a small window size, especially when containers have long names etc. 

**Describe the solution you'd like**
Adding a config option to allow shortening the container status from "running" to "R" for instance would save a lot of space while still being intuitive, the health info coule be similarly shortened too from "(healthy)" to (H).

**Describe alternatives you've considered**
Could also use suitable Unicode emoji to show the various states and healths, this could be included as a 3rd option.

**Comment by jesseduffield** [comment]
Makes sense to me. The possible states are: 
```
exited
created
running
paused
dead
restarting
removing
```
So using letters is a little hard, especially given we have three states that start with 'R'. Perhaps we could think of some more semantic characters to use. E.g. running: '▶︎', restarting '⟳', paused: ⏾︎.

**Comment by jesseduffield** [comment]
the code for this is over in `pkg/gui/presentation/containers.go` in the `getContainerDisplayStatus` function. We could create a map e.g.:
```go
var statusMap = map[string]string{
	"paused":     "⏾︎",
	"exited":     "x",
	"created":    "⏻︎",
	"removing":   "♲",
	"restarting": "⟳",
	"running":    "▶︎",
	"dead":       "!",
}
```
and use that there. But testing that out locally I see there's some glitchiness in the rendering which I believe comes from the gocui package (which is something we control, via a fork). I suspect the draw function in `vendor/github.com/jesseduffield/gocui/view.go` is the culprit.

**Comment by martini1992** [comment]
Using these symbols is highly dependant on the font used in the terminal and also support in the terminal itself which could cause issues. Even in the browser (Firefox on Windows) the chosen paused and created icons won't display, this is why I suggested normal characters. For the less common double ups, restarting and removing could we use "RS" and "RM"?

**Comment by jesseduffield** [comment]
two characters sounds like a good compromise

**Comment by jesseduffield** [comment]
I'm happy to just have either full text or the two characters (given my testing with single unicode characters proved buggy anyway).

And yep, it's just healthy, unhealthy, and starting. One character for that sounds good

**Comment by jesseduffield** [comment]
Would you be up to raising a PR for this @martini1992 ?

**Comment by martini1992** [comment]
Awesome, if nobody has picked this up by this evening (after 4pm GMT) then I'll see what I can put together. Bare with me though I'm only a lowly C# dev 🤣

**Comment by martini1992** [comment]
Struggling to figure out how to pass the AppConfig into GetContainerDisplayStrings in pkg/gui/presentation/containers.go
Adding any reference to Gui in containers.go throws up undefined errors.

**Comment by jesseduffield** [comment]
you can add the config as a new argument to `GetContainerDisplayStrings` and then pass it in from the place it's called. That should do the trick

**Comment by martini1992** [comment]
Yeah, I was trying to pass the whole Gui object, but just passing the GuiConfig seems to be working.

23 Nov 2022 22:03:28 Jesse Duffield ***@***.***>:

**Comment by martini1992** [comment]
Have a working version now :) I was thinking again about the issues you saw and I am seeing too using the icons, found gocui has been abandoned by it's creator, though there seems to be an active fork [https://github.com/awesome-gocui/gocui](https://github.com/awesome-gocui/gocui) might be worth looking into, crucially it claims to have better wide character support, which I believe is the issue we were seeing.

**Comment by jesseduffield** [comment]
We actually use our own [fork](https://github.com/jesseduffield/gocui) (under my github account) which is _supposed_ to have better wide character support. I'm not sure how much it's diverged from awesome-gocui though (the guys who made that were early contributors of lazygit/lazydocker)

**Comment by martini1992** [comment]
There seems to be quite a difference in the number of commits after each forked from the main project. I fI was to try testing awesome-gocui in place of your fork, would it be a case of replacing the reference in `go.mod` and replacing it in every source files import?

**Comment by martini1992** [comment]
Right, firstly sorry it's been a while.
I've messed around with this quite a bit and discovered that it seems to a mix of things causing the issue.
- Primarily, the padding functions in util was using len() instead of runewidth.StringWidth(), moving to this helped a great deal.
- However, the package go-runewidth doesn't always come back with the correct result (for reasons outlined in several issues such as this one # though it looks like they are working on it.
- It seems quite difficult to ensure that the unicode text representation of any symbols is used instead of the emoji counterparts (which are the default) adding Variation Selector-15 after the offending rune doesn't seem to work for every rune.
- Additionally using Variation Selector-15 counts as a character with width of 1 using runewidth.StringWidth() which is an issue since it's a non-printable character.

I have found a sweet spot selecting symbols which don't have emoji variants (currently) and seem to be either handled properly by runewidth.StringWidth() or with play nicely with the 1 character padding, so runes which display as 2 width but are reported by runewidth.StringWidth() as being 1 character wide (absorbing that 1 char space).

See what you think, new config is documented.
